perf(batch-exports): Use `distributed_events_recent` if backfill is within last 6 days

rossgray opened the pull request:
Currently we always query `events` for backfills, and use the `timestamp` column as the filter. However, we use `events_recent` for regular runs, with `inserted_at` as the filter. This means if a run fails or needs to be re-run for some reason, if we run a backfill, it won't run the same query as it originally did, which could lead to discrepancies. Since we have event data in `events_recent` for the last 7 days, it would make sense to use this table and filter on `inserted_at` when backfilling recent data.

## Changes

### Main logical change

Use `distributed_events_recent` for backfills that are entirely within the last 6 days (we choose 6 days to give ourselves a buffer to ensure we're not running into issues with deleted events). We want to consider the _entire_ backfill duration as otherwise this could lead to unexpected behaviour where the backfill is running different queries depending on which interval it's running.

### Consequential changes

We need to know the backfill start and end dates in addition to the individual interval start and end dates.

To handle this, and to think ahead to when we'll also need to accept a `backfill_id` so we can store this against an individual run, I'm creating a `BackfillDetails` dataclass that is added to the inputs.

Therefore, we need to update the workflow inputs to accept these new dates. In theory this should be straightforward, however we also need to consider backfills that are currently in progress.

For continuous exports that are already set up in Temporal everything should continue to work fine, as `backfill_details` is optional and defaults to `None`.

The one place I think it could be an issue is for backfills that are in progress at the time of the deployment: The backfill workflow will be creating workflows with the old inputs but the workers could be running the new code which is not expecting the old inputs.
Therefore I need to include both the old and new inputs to all the workflows to handle this migration: the workflows should accept `is_backfill`, `is_earliest_backfill` and also `backfill_details`, taking the inputs first from `backfill_details` if present, otherwise defaulting to the old arguments.

Once we're confident all existing backfills have completed we can remove the old arguments

### Cleanup/refactoring

I've refactored the workflow/activity inputs to inherit from a base class, so it is clear which inputs are common across all batch exports and which are specific to the destination.

I'm removing the previous `BATCH_EXPORT_DISTRIBUTED_EVENTS_RECENT_ROLLOUT` setting since this has already been rolled out to 100% of teams

### Testing

Since we're now using `distributed_events_recent` (or `events_recent`) for the majority of our batch exports (and for all the realtime batch exports) it means a lot of tests were failing because the fixtures are inserting into `sharded_events`.

I think it makes sense for our tests to now use `events_recent` by default. This requires [updating the default `data_interval_end` to be the current day](https://github.com/PostHog/posthog/pull/28323/files#diff-3c97a7e281adf43587104003f8de324f11daea4f5eae9e730ef9b78be52bf988R278).

I also [added a test](https://github.com/PostHog/posthog/pull/28323/files#diff-edeee6f7aa5bc8c2ef2b97cdf49a8f9a21dec563b572a81d7a5906726c0727b8R2175) to `test_s3_batch_export_workflow.py` to assert batch exports are selecting from the expected table, since the logic is becoming quite complex.

## Does this work well for both Cloud and self-hosted?

Should do.

## How did you test this code?

Automated tests

rossgray committed:
Refactor inputs

rossgray committed:
Update inputs across all BE destinations

Pass into spmc

Add debug logging for migration

Update legacy batch exports

Update tests

rossgray committed:
Pass in backfill_details to producer

rossgray committed:
WIP

rossgray committed:
Add a new test to test query

rossgray committed:
Use distributed_events_recent table unless backfill in past

rossgray committed:
Use events_recent by default for tests

rossgray committed:
Fix tests

rossgray committed:
Fix more tests

rossgray committed:
Fix more tests

rossgray committed:
Fix more tests

rossgray committed:
Fix more tests

rossgray committed:
Use docstring

tomasfarias commented on posthog/batch_exports/service.py:
```diff
@@ -930,3 +926,37 @@ def fetch_earliest_backfill_start_at(
+    exclude_events: list[str] | None = None
+    include_events: list[str] | None = None
+    run_id: str | None = None
+    # TODO: Remove after updating existing batch exports to use backfill_details
+    is_backfill: bool = False
+    backfill_details: BackfillDetails | None = None
+    batch_export_model: BatchExportModel | None = None
+    # TODO: Remove after updating existing batch exports
+    batch_export_schema: BatchExportSchema | None = None
```
Just a small nit, let's put all the "to delete" stuff at the bottom: 
```suggestion
    backfill_details: BackfillDetails | None = None
    batch_export_model: BatchExportModel | None = None
    # TODO: Remove after updating existing batch exports
    batch_export_schema: BatchExportSchema | None = None
    # TODO: Remove after updating existing batch exports to use backfill_details
    is_backfill: bool = False
```

tomasfarias commented on posthog/temporal/batch_exports/batch_exports.py:
```diff
@@ -170,14 +177,18 @@ async def iter_model_records(
     interval_start: str | None,
     interval_end: str,
     destination_default_fields: list[BatchExportField] | None = None,
     # TODO - remove this once all batch exports are using the latest schema
     use_latest_schema: bool = False,
+    # TODO: this can be removed once all backfill inputs are migrated
+    is_backfill: bool = False,
+    backfill_details: BackfillDetails | None = None,
```
Similar nitpick as before, to make it clear what is to delete, move `backfill_details` up: 
```suggestion
    backfill_details: BackfillDetails | None = None,
    # TODO - remove this once all batch exports are using the latest schema
    use_latest_schema: bool = False,
    # TODO: this can be removed once all backfill inputs are migrated
    is_backfill: bool = False,
```

tomasfarias commented on posthog/temporal/batch_exports/batch_exports.py:
```diff
@@ -241,6 +255,9 @@ async def iter_records_from_model_view(
         filters_str = f"AND {filters_str}"
 
+    # TODO: this can be simplified once all backfill inputs are migrated
+    is_backfill = (backfill_details is not None) or is_backfill
```
In case we refactor this before removing it, I'd rather call `get_is_backfill` here so that we would only have to refactor one place:
```
```suggestion
    is_backfill = backfill_details.get_is_backfill()
```

tomasfarias commented on posthog/temporal/batch_exports/spmc.py:
```diff
@@ -580,6 +596,9 @@ async def start(
         else:
             filters_str, extra_query_parameters = "", extra_query_parameters
 
+        # TODO: this can be simplified once all backfill inputs are migrated
+        is_backfill = (backfill_details is not None) or is_backfill
+
```
```suggestion
        # TODO: this can be simplified once all backfill inputs are migrated
        is_backfill = backfill_details.get_is_backfill()

```

tomasfarias commented on posthog/temporal/tests/batch_exports/conftest.py:
```diff
@@ -232,12 +232,16 @@ def data_interval_start(request, data_interval_end, interval):
     except AttributeError:
         pass
-    return dt.datetime(2023, 4, 25, 15, 0, 0, tzinfo=dt.UTC)
+    return dt.datetime.now(tz=dt.UTC).replace(hour=0, minute=0, second=0, microsecond=0)
```
Have you ran the bigquery test suite? We use freezegun and do a lot of `with freeze_time(TEST_TIME)` with a hardcoded `TEST_TIME`, so I'm wondering if this call is impacted in any way.

tomasfarias commented on posthog/temporal/tests/batch_exports/test_postgres_batch_export_workflow.py:
```diff
@@ -659,7 +663,7 @@ async def test_postgres_export_workflow_without_events(
     # This is hardcoded relative to the `data_interval_end` used in all or most tests, since that's also
     # passed to `generate_test_data` to determine the timestamp for the generated data.
-    [dt.datetime(2023, 4, 24, 15, 0, 0, tzinfo=dt.UTC)],
+    [dt.datetime.now(tz=dt.UTC).replace(hour=0, minute=0, second=0, microsecond=0) - dt.timedelta(hours=24)],
     indirect=True,
 )
 @pytest.mark.parametrize("interval", ["hour"], indirect=True)
```
Shouldn't this be `- dt.timedelta(hours=1)` given that we are only testing hourly?

rossgray replied:
from my understanding, this test is testing when data exists outside the data interval (see below on line 683):
> We expect persons outside the batch interval to also be backfilled (i.e. persons that were updated more than an hour ago) when setting `is_earliest_backfill=True`.

Also the data interval in this test was previously 24 hours: `data_interval_start` was  `dt.datetime(2023, 4, 24, 15, 0, 0, tzinfo=dt.UTC)` and `data_interval_end` was by default `dt.datetime(2023, 4, 25, 15, 0, 0, tzinfo=dt.UTC`

tomasfarias commented on posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:
```diff
@@ -972,7 +980,7 @@ async def test_s3_export_workflow_with_minio_bucket(
     # This is hardcoded relative to the `data_interval_end` used in all or most tests, since that's also
     # passed to `generate_test_data` to determine the timestamp for the generated data.
-    [dt.datetime(2023, 4, 24, 15, 0, 0, tzinfo=dt.UTC)],
+    [dt.datetime.now(tz=dt.UTC).replace(hour=0, minute=0, second=0, microsecond=0) - dt.timedelta(hours=24)],
```
Same as previous comment, why not `- dt.timedelta(hours=1)`?

rossgray replied:
as above

tomasfarias commented on posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:
```diff
@@ -2199,29 +2229,103 @@ async def test_insert_into_s3_activity_when_using_distributed_events_recent_tabl
+            # Set up the mock to return our async iterator
+            self.mock_client.astream_query_as_arrow.return_value = self._create_record_batch_iterator()
+
+        def expect_query_string(self, expected_query_string: str) -> None:
+            """Assert that the executed query contains the expected string."""
+            assert self.mock_client.astream_query_as_arrow.call_count == 1
+            call_args = self.mock_client.astream_query_as_arrow.call_args
+            query = call_args[0][0]  # First positional argument of the first call
+            assert expected_query_string in query
```
Doesn't look like we are calling this method anywhere?

rossgray replied:
yes, you're right, I forgot to delete this in a previous refactor. Will clean up 👍

rossgray committed:
Update ordering of input args

Co-authored-by: Tomás Farías Santana <tomas@tomasfarias.dev>

tomasfarias reviewed (approved):
Changes are a net positive, also great work on tests. I've made a few comments to address. Mostly nits, but also a couple ones to address. Should be fine to merge this after those are solved.

rossgray commented on posthog/temporal/batch_exports/batch_exports.py:
```diff
@@ -492,6 +498,9 @@ def iter_records(
     start_at = dt.datetime.fromisoformat(interval_start) if interval_start is not None else None
     end_at = dt.datetime.fromisoformat(interval_end)
 
+    # TODO: this can be simplified once all backfill inputs are migrated
+    is_backfill = (backfill_details is not None) or is_backfill
+
```
as above

rossgray committed:
cleanup

rossgray commented:
Thanks for the review @tomasfarias ! I've addressed all the feedback (either making the suggested changes or replying to the comments) so should be good for a final re-review

tomasfarias reviewed (approved):
All good from me :+1: There is a red test though, but besides that I'm good to merge this!
